python/hidet/ffi/runtime_api.py

```diff
@@ -9,11 +9,43 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from typing import Union
-from ctypes import c_void_p, c_char_p, c_uint64, c_int32
+from typing import Union, Optional, List
+from ctypes import c_void_p, c_char_p, c_uint64, c_int32, c_int, pointer, Structure, c_byte, POINTER
+from enum import IntEnum
 from hidet.cuda import Stream
-from .ffi import get_func
+from .ffi import get_func, nccl_available
 
+class NcclUniqueId(Structure):
+    """
+    Defined in nccl.h
+    """
+    _fields_ = [("internal", c_byte * 128)]
+
+class NcclCommunicator:
+    """
+
+    """
+    def __init__(self, handle: int):
+        """
+        Users should not call this constructor directly. Because there are two ways of creating
+        a new communicator: 1) using unique_id and rank ; 2) using split.
+        """
+        if not nccl_available():
+            raise RuntimeError("NCCL Library not found.")
+        self._handle = handle
+
+        # TODO: how to ensure the following two are identical?
+        _comms.append(self)
+        runtime_api.add_nccl_comm(self)
+    
+    def __del__(self):
+        """
+        Should we manage the lifetime of communicator object in Python or C++?
+        """
+        nccl_runtime_api.comm_destroy(self)
+
+    def split(self):
+        raise NotImplementedError()
```

Comment:
Consider moving this to "hidet/cuda/nccl.py".

Reply:
Good questions!

> How should we manage the lifetime of NcclCommunicator objects in Python and ncclComm_t objects in C++?

Because different flow graphs may have different set of communicators, thus, we should create the communicators for each flow graph. Let's keep the communicators in CompiledGraph. Something like

```python
class CompiledGraph:
def __init__(...):
    ...
    self.nccl_comms = ... # create the communicators, 
                          # let's store it as hidet.ffi.utils.Array(void*, num of comm)
                          # and create a runtime api called 
                          # "void set_nccl_comms(int num_comm, void** comm_array)"
def run_async(...):
    ...
    runtime_api.set_nccl_comms(len(self.nccl_comms), self.nccl_comms)
    ...

def __del__(...):
    # destroy communicators
```

For `FlowGraph.forward(...)`, we can raise an error for now if we find it is a distributed flow graph. 

> Should we import 'nccl.py' in 'nccl/ffi.py' or the opposite? Because NcclCommunicator might depend on some runtime api, therefore we need 'nccl/ffi.py' in 'nccl.py' which might cause circular import.

Consider: let the nccl runtime api return integer (the pointer to communicator), and the `hidet/cuda/nccl/nccl.py` imports `hidet/cuda/nccl/ffi.py`. The nccl ffi should only be used by `nccl.py` and we expose the api to the users in `nccl.py`.